File: sandbox-runner/patterns/retry-metrics/python/starter.py
import asyncio
import time

from temporalio.client import Client

from shared import ALERT_THRESHOLD, TASK_QUEUE, WORKFLOW_ID_PREFIX
from workflows import MonitoredRetryWorkflow


async def main() -> None:
    client = await Client.connect("localhost:7233")
    run_id = f"{WORKFLOW_ID_PREFIX}-{int(time.time() * 1000)}"

    print("Starting monitored retry workflow.")
    print(f"  Alert threshold: attempt > {ALERT_THRESHOLD}")
    print("  Service recovers on attempt 6 — metric will be emitted on attempts 4 and 5.")
    print()

    handle = await client.start_workflow(
        MonitoredRetryWorkflow.run,
        "https://api.example.com/data",
        id=run_id,
        task_queue=TASK_QUEUE,
    )
    print(f"Started workflow: {handle.id}")

    result = await handle.result()
    print(f"Workflow result: {result}")
    print()
    print(f"Open the Temporal UI and search for '{handle.id}' to inspect the retry history.")


if __name__ == "__main__":
    asyncio.run(main())

File: sandbox-runner/patterns/retry-metrics/typescript/starter.ts
import { Client, Connection } from "@temporalio/client";

import { ALERT_THRESHOLD, TASK_QUEUE, WORKFLOW_ID_PREFIX } from "./shared";
import { monitoredRetryWorkflow } from "./workflows";

async function main(): Promise<void> {
  const connection = await Connection.connect();
  try {
    const client = new Client({ connection });
    const runId = `${WORKFLOW_ID_PREFIX}-${Date.now()}`;

    console.log("Starting monitored retry workflow.");
    console.log(`  Alert threshold: attempt > ${ALERT_THRESHOLD}`);
    console.log(
      "  Service recovers on attempt 6 — metric will be emitted on attempts 4 and 5.",
    );
    console.log();

    const handle = await client.workflow.start(monitoredRetryWorkflow, {
      args: ["https://api.example.com/data"],
      taskQueue: TASK_QUEUE,
      workflowId: runId,
    });
    console.log(`Started workflow: ${handle.workflowId}`);

    const result = await handle.result();
    console.log(`Workflow result: ${result}`);
    console.log();
    console.log(
      `Open the Temporal UI and search for '${handle.workflowId}' to inspect the retry history.`,
    );
  } finally {
    await connection.close();
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
